test/topology_custom/test_mv_read_concurrency.py

```diff
@@ -0,0 +1,82 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+from test.pylib.manager_client import ManagerClient
+
+import asyncio
+import pytest
+import logging
+
+from test.topology.conftest import skip_mode
+from test.pylib.util import wait_for_view
+from cassandra import ReadTimeout, WriteTimeout
+
+logger = logging.getLogger(__name__)
+
+# This test verifies that the writes causing view updates don't impact the latency of regular reads
+# due to contention on the read concurrency semaphore.
+# The test creates a table and a table with a materialized view, and then runs a large number of writes causing view updates
+# while concurrently running a small read workload on the other table.
+# The test fails if any of the reads times out.
+# Reproduces https://github.com/scylladb/scylladb/issues/8873
+@pytest.mark.asyncio
+@skip_mode('release', "error injections aren't enabled in release mode")
+async def test_mv_read_concurrency(manager: ManagerClient) -> None:
+    node_count = 1
+    # Disable cache to make reads use the read concurrency semaphore.
+    # Tests remove the rcs multiplier by default, here we use a slightly smaller one (1 instead of default 2) to hit the issue faster.
+    servers = await manager.servers_add(node_count, config={'enable_tablets': True, 'enable_cache': False, 'reader_concurrency_semaphore_serialize_limit_multiplier': 1})
+
+    cql, _ = await manager.get_ready_cql(servers)
+    await cql.run_async(f"CREATE KEYSPACE ks WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': 1}}"
+                        "AND tablets = {'initial': 2}")
+    await cql.run_async(f"CREATE TABLE ks.tab (p int PRIMARY KEY, mvp int, v text)")
+    await cql.run_async(f"CREATE TABLE ks.tab2 (p int PRIMARY KEY, mvp int)")
+    await cql.run_async(f"CREATE MATERIALIZED VIEW IF NOT EXISTS ks.mv AS SELECT p, mvp FROM ks.tab \
+        WHERE p IS NOT NULL AND mvp IS NOT NULL PRIMARY KEY (mvp, p)")
+    await wait_for_view(cql, 'mv', node_count)
+
+    row_count = 300
+    for i in range(10):
+        await cql.run_async(f"INSERT INTO ks.tab2 (p, mvp) VALUES ({i}, {i})")
+
+    # The injection prolongs the time we hold the read concurrency semaphore resources during the rbw during a view update
+    await manager.api.enable_injection(servers[0].ip_addr, "keep_mv_read_semaphore_units", one_shot=False)
+
+    failed = None
+    stop_event = asyncio.Event()
+    async def do_read(i: int):
+        read_stmt = cql.prepare(f"SELECT mvp FROM ks.tab2 WHERE p=? USING TIMEOUT 10s")
```

Comment:
So before your fix, this test failed on these 10 second timeouts, and your patch fixes it? Nice.

Reply:
Right